pkg/agent/core/ngt/service/ngt.go

```diff
@@ -670,16 +682,22 @@ func (n *ngt) saveIndex(ctx context.Context) (err error) {
 		return
 	}
 	atomic.SwapUint64(&n.lastNoice, noice)
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-	// wait for not indexing & not saving
-	for n.IsIndexing() || n.IsSaving() {
-		runtime.Gosched()
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-ticker.C:
+	err = func() error {
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		// wait for not indexing & not saving
+		for n.IsIndexing() || n.IsSaving() {
+			runtime.Gosched()
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-ticker.C:
+			}
 		}
+		return nil
+	}()
+	if err != nil {
+		return err
```

Comment:
should we put lock here also?

Reply:
why???

Reply:
same as createIndex, to avoid multiple goroutine to execute saveIndex simultaneously

Reply:
@kevindiu 
To avoid that, I think `saveIndex` has the following code.
If called multiple times in multiple goroutine, `noice` will be the same as`lastNoice` except for the first time, so `saveIndex` may not be executed.

https://github.com/vdaas/vald/blob/b9e90f944715c488f76daf1a7878f08303835863/pkg/agent/core/ngt/service/ngt.go#L680-L684